### ISSUE TITLE
Custom Plugin Assistance

### ISSUE THREAD

**Issue by sereeds** [firehol/netdata#3512]
I am trying to create a proper plugin for use with Netdata.  I currently have a shell script that just sends the data to Netdata's with netcat.  However, I now see there are much better ways to achieve this, the most ideal being a python script plugin.

I took a look at the custom plugin examples for both bash and python, but they are beyond my expertise.  I was wondering if someone could assist me, and turn this shell script into a proper python plugin for netdata.

I am trying to collect some data from the OpenVPN status log file, similar to what the existing openvpnstatus does, but extending it further.
I need to read 4 OpenVPN status files, and iterate over the client connections.
Then I can collect and graph the uptime, and possibly bytes transferred of the clients I choose.
So my shell command is this:
  cat vpn[1-4].sts | grep client1
That will specifically check to see if client1 is connected.
If there is no result, then the client is not connected.  If a result, then we know client1 is connected on one of those VPNs, at the time that I read the status log files.  So 1 uptime graph for each client.
I am updating the OpenVPN status files every 5 seconds, so that can also be my netdata rate for these graphs.
Ideally, in the plugin configuration we could specify 1 or many OpenVPN status log files, and then also specify 1 or many clients to monitor and graph.  No need for graphing all clients, I would like to specify individual ones to monitor.
I will likely use badges instead of graphs, but the idea is to collect the data in a more appropriate way for netdata, python plugin.

Does that make sense?  I apologize if I don't explain it well.  

I was hoping someone with more python skills than I, might say, "oh ya that's easy to make".
Maybe we can just extend the existing openvpnstatus plugin?
Any help is appreciated.

**Comment by l2isbad** [comment]
Actually a good idea about a client uptime.
I see ovpn status log file has `Connected Since` for connected users.
So we can chart [uptime](http://london.my-netdata.io/default.html#menu_system_submenu_uptime;theme=slate;help=true).

**Comment by l2isbad** [comment]
@sereeds 
Do you mean "connection status" (connected or not) by uptime in your request?

**Comment by sereeds** [comment]
I was looking to collect the "connection status" for the client, from the status file.  I used the example shell script plugin, with this script:

my "get" funtion:

	test_rc108="$(/usr/bin/grep -E ^rc108 /home/vpn3.sts)"
	if [ -z "$test_rc108" ]
		then
		test_rc108=0
	else
		test_rc108=1
	fi

the create funtion:

CHART test.rc108 '' "RC108" "Connected" connection connection line $((test_priority)) $test_update_every
DIMENSION connected '' absolute 1 1


So I am just basically recording a 0 or 1 for connected status.  However, I see that status file does also show "Connected Since", so we may as well record that metric too right?

My silly shell script example only monitors that one client.  The ideal plugin would iterate over a list of VPN clients that I want to chart, specified in the plugin config.  

Also note, there are different types of OpenVPN status files.  The Type 2 provides more a little more data on the "Connected Since".  Some distributions force it to status type 2 in the systemd unit file, which I had to remove to get back to status file type 1.
Here are a couple snipplets to show the difference.

Type 1:
OpenVPN CLIENT LIST
Updated,Thu Mar  8 07:06:54 2018
Common Name,Real Address,Bytes Received,Bytes Sent,Connected Since
rc108,192.168.236.3:53516,273994,269992,Wed Mar  7 16:36:41 2018
rc140,192.168.79:59254,609539,520891,Wed Mar  7 16:36:48 2018
rc302,192.168.232.3:33479,119952,121185,Thu Mar  8 03:15:53 2018

Type 2:
TITLE,OpenVPN 2.4.4 x86_64-redhat-linux-gnu [Fedora EPEL patched] [SSL (OpenSSL)] [LZO] [LZ4] [EPOLL] [PKCS11] [MH/PKTINFO] [AEAD] built on Sep 26 2017
TIME,Fri Mar  9 07:20:13 2018,1520605213
HEADER,CLIENT_LIST,Common Name,Real Address,Virtual Address,Virtual IPv6 Address,Bytes Received,Bytes Sent,Connected Since,Connected Since (time_t),Username,Client ID,Peer ID
CLIENT_LIST,rc357,192.168.77.18:42093,10.41.78.113,,6134702,6033984,Thu Mar  1 01:10:12 2018,1519891812,UNDEF,52,0
CLIENT_LIST,test1,192.168.131.79:51848,10.41.78.131,,838777,800652,Wed Mar  7 16:36:31 2018,1520465791,UNDEF,88,3
CLIENT_LIST,rc101,192.168.78.39:23598,10.41.78.134,,6146698,5870324,Fri Feb 23 12:01:07 2018,1519412467,UNDEF,4,4
CLIENT_LIST,rc107,192.168.78.122:21360,10.41.78.133,,2433991,2447285,Sat Mar  3 12:55:12 2018,1520106912,UNDEF,69,5

The existing openvpn status log plugin, and my shell script, only work with status type 1.

Thank you :-)
